### PR TITLE
feat(EstimatorReport): Display the mean decrease impurity

### LINKED ISSUE #1323: Feat(EstimatorReport): Display the feature importance based on mean decrease impurity
### Is your feature request related to a problem? Please describe.

As a Data Scientist, to explain my model and understand the problem I'm trying to solve, I need to check the feature importance by a permutation method. This should be available to tree models only.

### Describe the solution you'd like

```python
df = report.feature_importance.mean_decrease_impurity() # renders a dataframe
display = report.feature_importance.plot.mean_decrease_impurity() # renders a display
```

### Describe alternatives you've considered, if relevant

**API remark**  
Later, if the object report contains too many accessors, we will group the feature importance and add a parameter to decide which of the feature importance type we want to display.

**technical remark**  
This is a nice method because it's very easily accessible by `forest.feature_importances_`. However, it's computed on the train set automatically. We might want to do it on the test set. For now, this will not be an option.  

### Additional context

part of the epic https://github.com/probabl-ai/skore/issues/1314

### PR DESCRIPTION
**Pull request by auguste-probabl**
Closes #1323 

Todo:

- [x] Add tests on all sklearn estimators
- [ ] Add example
- [x] Add API docs

### CODE REVIEW

**Review by MarieSacksick — COMMENTED** [review]
Tested, and appart from the missing example (which will be done by enriching the current one in progress I guess?), it's good for me :) !

File: skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    from sklearn.preprocessing import PolynomialFeatures
+
+    X, y = regression_data
+    X = pd.DataFrame(X, columns=[f"my_feature_{i}" for i in range(5)])
+
+    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
+
+    model = make_pipeline(
+        PolynomialFeatures(degree=2, interaction_only=True),
+        RandomForestRegressor(random_state=0),
```
Comment (glemaitre) [inline skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py:191]:
```suggestion
        RandomForestRegressor(n_estimators=2, random_state=0),
```

File: skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+            sklearn.ensemble.ExtraTreesRegressor(),
+            sklearn.ensemble.GradientBoostingClassifier(),
+            sklearn.ensemble.GradientBoostingRegressor(),
+            sklearn.ensemble.RandomForestClassifier(),
+            sklearn.ensemble.RandomForestRegressor(),
+            sklearn.ensemble.RandomTreesEmbedding(),
+            sklearn.tree.DecisionTreeClassifier(),
+            sklearn.tree.DecisionTreeRegressor(),
+            sklearn.tree.ExtraTreeClassifier(),
+            sklearn.tree.ExtraTreeRegressor(),
```
Comment (glemaitre) [inline skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py:149]: To keep the execution time quick since we are interested in testing the statistical soundness.

```suggestion
            sklearn.ensemble.AdaBoostClassifier(n_estimators=2),
            sklearn.ensemble.AdaBoostRegressor(n_estimators=2),
            sklearn.ensemble.ExtraTreesClassifier(n_estimators=2),
            sklearn.ensemble.ExtraTreesRegressor(n_estimators=2),
            sklearn.ensemble.GradientBoostingClassifier(n_estimators=2),
            sklearn.ensemble.GradientBoostingRegressor(n_estimators=2),
            sklearn.ensemble.RandomForestClassifier(n_estimators=2),
            sklearn.ensemble.RandomForestRegressor(n_estimators=2),
            sklearn.ensemble.RandomTreesEmbedding(n_estimators=2),
            sklearn.tree.DecisionTreeClassifier(),
            sklearn.tree.DecisionTreeRegressor(),
            sklearn.tree.ExtraTreeClassifier(),
            sklearn.tree.ExtraTreeRegressor(),
```

File: skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    assert result.index.tolist() == expected_index
+
+    expected_columns = ["Mean decrease impurity"]
+    assert result.columns.tolist() == expected_columns
+
+
+@pytest.mark.parametrize(
+    "estimator",
+    [
+        RandomForestClassifier(random_state=0),
```
Comment (glemaitre) [inline skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py:104]:
```suggestion
        RandomForestClassifier(n_estimators=2, random_state=0),
```

File: skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+            (5, 1),
+        ),
+        (
+            make_classification(n_features=5, random_state=42),
+            make_pipeline(StandardScaler(), RandomForestClassifier(random_state=0)),
+            (5, 1),
+        ),
+        (
+            make_regression(n_features=5, n_targets=3, random_state=42),
+            RandomForestRegressor(random_state=0),
```
Comment (glemaitre) [inline skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py:77]:
```suggestion
            RandomForestRegressor(n_estimators=2, random_state=0),
```

File: skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+                n_samples=30,
+                n_informative=3,
+                random_state=42,
+            ),
+            make_pipeline(StandardScaler(), RandomForestClassifier(random_state=0)),
+            (5, 1),
+        ),
+        (
+            make_classification(n_features=5, random_state=42),
+            make_pipeline(StandardScaler(), RandomForestClassifier(random_state=0)),
```
Comment (glemaitre) [inline skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py:72]:
```suggestion
            make_pipeline(StandardScaler(), RandomForestClassifier(n_estimators=2, random_state=0)),
```

File: skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        ),
+        (
+            make_classification(
+                n_features=5,
+                n_classes=3,
+                n_samples=30,
+                n_informative=3,
+                random_state=42,
+            ),
+            make_pipeline(StandardScaler(), RandomForestClassifier(random_state=0)),
```
Comment (glemaitre) [inline skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py:67]:
```suggestion
            make_pipeline(StandardScaler(), RandomForestClassifier(n_estimators=2, random_state=0)),
```

File: skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+        ),
+        (
+            make_classification(
+                n_features=5,
+                n_classes=3,
+                n_samples=30,
+                n_informative=3,
+                random_state=42,
+            ),
+            RandomForestClassifier(random_state=0),
```
Comment (glemaitre) [inline skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py:56]:
```suggestion
            RandomForestClassifier(n_estimators=2, random_state=0),
```

File: skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+    "data, estimator, expected_shape",
+    [
+        (
+            make_classification(n_features=5, random_state=42),
+            RandomForestClassifier(random_state=0),
+            (5, 1),
+        ),
+        (
+            make_classification(n_features=5, random_state=42),
+            RandomForestClassifier(random_state=0),
```
Comment (glemaitre) [inline skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py:45]:
```suggestion
            RandomForestClassifier(n_estimators=2, random_state=0),
```

File: skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+    ]
+    assert result.columns.tolist() == ["Mean decrease impurity"]
+
+
+@pytest.mark.parametrize(
+    "data, estimator, expected_shape",
+    [
+        (
+            make_classification(n_features=5, random_state=42),
+            RandomForestClassifier(random_state=0),
```
Comment (glemaitre) [inline skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py:40]:
```suggestion
            RandomForestClassifier(n_estimators=2, random_state=0),
```

File: skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+from sklearn.pipeline import Pipeline, make_pipeline
+from sklearn.preprocessing import StandardScaler
+from skore import EstimatorReport
+
+
+def test(classification_data):
+    X, y = classification_data
+    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)
+
+    forest = RandomForestClassifier(random_state=0)
```
Comment (glemaitre) [inline skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py:17]:
```suggestion
    forest = RandomForestClassifier(n_testimators=2, random_state=0)
```

File: skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+from sklearn.base import is_regressor
+from sklearn.datasets import make_classification, make_regression
+from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
+from sklearn.model_selection import train_test_split
+from sklearn.pipeline import Pipeline, make_pipeline
+from sklearn.preprocessing import StandardScaler
+from skore import EstimatorReport
+
+
+def test(classification_data):
```
Comment (glemaitre) [inline skore/tests/unit/sklearn/estimator/feature_importance/test_mean_decrease_impurity.py:13]: Is this test already part of the test below? We could remove it.

**Review by glemaitre — COMMENTED** [review]
This PR looks good to me. Only a couple of changes to speed-up the test. In terms of documentation, I think that we should be amending the example created in the linear model PR.

File: skore/src/skore/sklearn/_estimator/feature_importance_accessor.py [hunk 12]
```diff
@@ -72,27 +72,31 @@ def coefficients(self) -> pd.DataFrame:
...
-        linear_model = estimator[-1] if isinstance(estimator, Pipeline) else estimator
-        intercept = np.atleast_2d(linear_model.intercept_)
-        coef = np.atleast_2d(linear_model.coef_)
+        estimator = (
+            parent_estimator[-1]
+            if isinstance(parent_estimator, Pipeline)
+            else parent_estimator
+        )
+        intercept = np.atleast_2d(estimator.intercept_)
+        coef = np.atleast_2d(estimator.coef_)
```
[the remark this answers is not in the document]
Comment (auguste-probabl) [reply skore/src/skore/sklearn/_estimator/feature_importance_accessor.py:95]: i'd rather keep it here to avoid delaying 1339

**Review by glemaitre — APPROVED** [review]
To ease the integration of the documentation, I think this PR can be merged. I can have another look in this week to see if we should have subsequent works.
